### PR TITLE
Expand the libsailfishapp documentation

### PR DESCRIPTION
**Pull request by llewelld**
Includes the following changes.
1. Adds the lib source files (.cpp, .h) to the doc generation process.
2. Expands the documentation sections with explanatory text and example
code (removing the TODOs in the process).
3. Adds brief documentation for each of the functions in the SailfishApp
namespace.

### CODE REVIEW

File: src/sailfishapp.cpp [hunk 1]
```diff
@@ -40,10 +40,73 @@
...
 namespace SailfishApp
 {
 
+/*!
+  \fn QGuiApplication * SailfishApp::application(int &argc, char **argv)
+
+  \brief Create a new QGuiApplication instance configured for use on SailfishOS
+
+  Creates and configures a QGuiApplication ready to be used as part of a
+  SailfishOS application.
```
Comment (martyone) [inline src/sailfishapp.cpp]: Should be maybe "Sailfish application", but definitely "Sailfish OS" instead of "SailfishOS". I am not fully sure about the guidelines, so rather check with someone more involved in the UX design :) Applies to all occurrences of "Sailfish".

**Comment by llewelld** [comment]
Comment from @vigejolla, coped over from elsewhere:

> LGTM, but I want to hear the opinions of martyone and Jare as well before approving. Also, please make sure that this ends up in the website as well (in addition to the sdk).

**Comment by vigejolla** [comment]
> Possibly replace this https://sailfishos.org/wiki/Tutorial_-_Combining_C%2B%2B_with_QML .

Definitely not replace. That tutorial gives instructions on how to create a model in C++ and how to use that in QML, which I would consider a slightly more advanced topic. The title of that page is just wrong and we should probably improve it. This documentation here covers more basic concepts. Both are useful documents.

**Comment by DylanVanAssche** [comment]
Great to see this improvement for third party devs! Thank you!

File: doc/index.qdoc [hunk 2]
```diff
@@ -1,68 +1,94 @@
...
+the \l SailfishApp configuration option. You might want to set this option if
 for some reason you intend to ship your QML files as Qt Resources (qrc).
 We recommend that you keep the default setup and install QML files in the
 filesystem instead.
 
-In case you provide application icon in multiple sizes:
+\section2 Deploying app icons in multiple sizes
+
+It's recommended that you provide application icons in multiple sizes, to
+support different device types and different screen dpi settings. You can do
```
[the remark this answers is not in the document]
Comment (llewelld) [reply doc/index.qdoc]: Thank you for spotting this; now fixed.

File: doc/index.qdoc [hunk 3]
```diff
@@ -1,68 +1,94 @@
...
 \li Add \c{SAILFISHAPP_ICONS = size1 size2...} to your .pro file,
-    enumerating all available icon sizes in form \c{WIDTHxHEIGHT}
+    enumerating all available icon sizes in the form \c{WIDTHxHEIGHT}
 \li Ensure all icons are present as \c{icons/WIDTHxHEIGHT/$$TARGET.png}
     in the source tree
 \endlist
 
-\section1 How to use i18n support in your project
+We recommend you include the following icons sizes in your application at a
+minimum: 86x86 108x108 128x128 172x172.
```
[the remark this answers is not in the document]
Comment (llewelld) [reply doc/index.qdoc]: Yes, good advice. I've made the change.

File: doc/index.qdoc [hunk 4]
```diff
@@ -92,81 +117,1022 @@ To add a new language to your project:
...
-Be sure to re-enable the line when you do the release build.
+in your qmake .pro file and the translations will not be built/updated. Be
+sure to re-enable the line when you do the release build.
 
-\section1 libsailfishapp C++ API
+\section1 SailfishApp C++ API
 
-TODO: Describe the API calls libsailfishapp provides
+The libsailfishapp C++ API can be used to quickly instantiate and configure an
+application that fits in to the Sailfish environment and way of operating.
```
[the remark this answers is not in the document]
Comment (llewelld) [reply doc/index.qdoc]: I've filled this out a bit and tried to give a brief summary of what the library actually does. Is this clearer now.

File: doc/index.qdoc [hunk 5]
```diff
@@ -92,81 +117,1022 @@ To add a new language to your project:
...
+\code
+#include <QtQuick>
+#include <QDebug>
+#include <sailfishapp.h>
+
+int main(int argc, char *argv[])
+{
+    QScopedPointer<QGuiApplication> app(SailfishApp::application(argc, argv));
+    QScopedPointer<QQuickView> view(SailfishApp::createView());
+    QQmlContext *ctxt = view->rootContext();
```
[the remark this answers is not in the document]
Comment (llewelld) [reply doc/index.qdoc]: Yes, good point. I've changed it to `context`.

File: doc/index.qdoc [hunk 6]
```diff
@@ -92,81 +117,1022 @@ To add a new language to your project:
...
+To reiterate, accessing QML objects from C++ is generally a bad idea. Where
+data must be shared between C++ and QML it's preferable to create objects with
+properties on the C++ side and expose them as context properties in the QML.
+
+Note that if you're planning to submit your application for inclusion in the
+Jolla Store, in order for it to pass \l {https://harbour.jolla.com/} {Harbour
+QA} the \c qmlRegisterType import must not match any of those listed in \l
+{https://github.com/sailfishos/sdk-harbour-rpmvalidator/blob/harbour-qa/disallowed_qmlimport_patterns.conf}
+{disallowed_qmlimport_patterns.conf}. See the \l
+{https://harbour.jolla.com/faq#QML_API} {Jolla Harbour FAQ} for more info.
```
[the remark this answers is not in the document]
Comment (llewelld) [reply doc/index.qdoc]: I also made this change.

File: src/sailfishapp.cpp (hunk 1 above)
Comment (llewelld) [reply src/sailfishapp.cpp]: I've gone through and changed references to "Sailfish OS" for the operating system and "Sailfish app" or "Sailfish application" for the application. Hopefully this is correct (I'll check). It is at least now consistent throughout.

**Comment by llewelld** [comment]
Thanks for the useful comments @martyone. I've fixed the easy stuff, but still need to move the code into imports and compile it as part of the build. That's a very good idea.

I've not made any changes to the structure, but will do so based on whatever is decided.

**Comment by llewelld** [comment]
@DylanVanAssche thanks for the comment, it's appreciated! Third party devs are the lifeblood.

**Comment by llewelld** [comment]
The latest change moves all of the code examples into projects that are built alongside the documentation. No RPM validation yet.... still working on that.

**Comment by martyone** [comment]
Nice, I think you could omit the packaging for most of the examples and keep it just for the example which demonstrates the overall app structure. Also the example code could be deduplicated and variants selected with #ifdefs and/or conditionally including files in one project.

**Comment by llewelld** [comment]
> Nice, I think you could omit the packaging for most of the examples and keep it just for the example which demonstrates the overall app structure. Also the example code could be deduplicated and variants selected with #ifdefs and/or conditionally including files in one project.

You're right, there's an obscene amount of duplication. I'm wondering whether it might be better (clearer for potential new devs looking at the code in the repo) to symlink the duplicated files rather than add conditionals to the project files. I'm thinking this would keep things clearer, while at the same time minimising the duplication. What do you think?

**Comment by martyone** [comment]
> You're right, there's an obscene amount of duplication. I'm wondering whether it might be better (clearer for potential new devs looking at the code in the repo) to symlink the duplicated files rather than add conditionals to the project files. I'm thinking this would keep things clearer, while at the same time minimising the duplication. What do you think?

I would divide the examples to two sets. 1) examples that demonstrate project and/or package structure and 2) examples that demonstrate programming techniques. In the former case I would first try to build an all-in-one example, so possibly without duplicates. The latter is likely to consist of brief enough snippets so that the reader should get good understanding just by reading the generated documentation without the need to explore the source code, so here I would choose such a solution that is more convenient for maintenance - conditionals IMO.

**Comment by llewelld** [comment]
@martyone, I've consolidate the projects, using the .pro files to avoid repeating resource usage and code as much as possible. All of the examples except three could be combined this way (and I left the first example as a stand-alone project, so that the interested newcomer can view a clean project structure).

I didn't go as far as adding conditionals into the source files as they're all rather short anyway, but I could do. What do you think?

File: examples/examples.pro [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    example04 \
+    example05 \
+    example06 \
+    example07 \
+    example08 \
+    example09 \
+    example10 \
+    example11 \
+    example12 \
+    example13
```
Comment (martyone) [inline examples/examples.pro]: You can directly enumerate the .pro files here

File: examples/example09/rpm/harbour-demo.spec [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+Name:       harbour-demo
+
+# >> macros
+# << macros
+
+%{!?qtc_qmake:%define qtc_qmake %qmake}
+%{!?qtc_qmake5:%define qtc_qmake5 %qmake5}
+%{!?qtc_make:%define qtc_make make}
+%{?qtc_builddir:%define _builddir %qtc_builddir}
```
Comment (martyone) [inline examples/example09/rpm/harbour-demo.spec]: These shouldn't be here anymore. Use "Builder: qmake5" in the yaml to get rid of them.

File: examples/example09/rpm/harbour-demo.spec [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+Summary:    harbour-demo
+Version:    0.1
+Release:    1
+Group:      Qt/Qt
+License:    LICENSE
+URL:        http://example.org/
+Source0:    %{name}-%{version}.tar.bz2
+Source100:  harbour-demo.yaml
```
Comment (martyone) [inline examples/example09/rpm/harbour-demo.spec]: I don't see the yamls in the commit.

File: examples/examples.pro (hunk 7 above)
Comment (llewelld) [reply examples/examples.pro]: Thanks, I've done so.

File: examples/example09/rpm/harbour-demo.spec (hunk 8 above)
Comment (llewelld) [reply examples/example09/rpm/harbour-demo.spec]: Thanks for this too, I've made this change and regenerated the spec files.

File: examples/example09/rpm/harbour-demo.spec (hunk 9 above)
Comment (llewelld) [reply examples/example09/rpm/harbour-demo.spec]: Yes, sorry, they should be there now.

**Comment by llewelld** [comment]
I've now split this documentation off into a separate repository: https://github.com/llewelld/sailfishapp-doc

If this structure is good, I'll clean this PR up to remove the general docs and just include the minor changes to the in-code documentation.

**Comment by llewelld** [comment]
I've removed (pretty much) all of the material from here as it's now in the [sdk-user-manual](https://github.com/sailfishos/sdk-user-manual). This PR now just includes some tidying of the original documentation, plus header documentation for the C++ code.

If this looks okay I'll squash the changes into a single commit before merging.

File: doc/index.qdoc [hunk 10]
```diff
@@ -1,68 +1,94 @@
...
+** Copyright (C) 2019 Jolla Ltd.
+** Contact: David Llewellyn-jones <david.llewellyn-jones@jollamobile.com>
+** All rights reserved.
+**
+****************************************************************************************/
+
 /*!
 \contentspage {Sailfish App Library}
-\page index.html
+\page index
```
Comment (vigejolla) [inline doc/index.qdoc]: Why did you remove ".html"? Now QtC complains that it can't find index.html!

Comment (llewelld) [reply doc/index.qdoc]: Good question. I don't recall why I changed this, but probably for no good reason. I've changed it back to 'index.html'.
